### PR TITLE
feat(composer): Support install before update and configure plugin behaviour

### LINKED ISSUE #8866: Renovate does not respect symfony/flex requirements
**What Renovate type, platform and version are you using?**

renovate/renovate:24.64.1-slim with Gitlab

**Describe the bug**

When you use Symfony, nowadays you also use `symfony/flex`, which has a feature to pin all Symfony packages within the same range (for example, 4.4.*). See https://symfonycasts.com/screencast/symfony5-upgrade/flex-symfony-require#extra-symfony-require for some explanation.

However, renovate does not respect this configuration as it does not have the flex plugin available when it asks Composer to update the lock file. And as flex is missing, it cannot limit the relevant `symfony/*` packages, which eventually results in extra, unwanted package updates (for optimal compatibility, you want to keep all symfony packages on the same version). 

This can be solved by running `composer install` before running `composer update`. That will make sure the flex plugin is available, and the subsequent install will adhere to the configured requirement by printing `Restricting packages listed in "symfony/symfony" to "4.4.*"` in the output.

I even believe this can have a positive impact on speed: install is generally fast (when you have a lock file), and the subsequent update will be faster as composer has less packages to consider (see the last part of the earlier symfony casts link).

**Relevant debug logs**

```
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":13,"level":20,"logContext":"zecDfDsSF","repository":"extern/007","branch":"renovate/drenso-symfony-shared-1.22.x","command":"docker run --rm --name=renovate_composer --label=renovate_child -v \"/builds/3cixCcwz/0/intern/renovate-runner/renovate/repos/gitlab/extern/007\":\"/builds/3cixCcwz/0/intern/renovate-runner/renovate/repos/gitlab/extern/007\" -v \"/builds/3cixCcwz/0/intern/renovate-runner/renovate/cache\":\"/builds/3cixCcwz/0/intern/renovate-runner/renovate/cache\" -e COMPOSER_CACHE_DIR -e COMPOSER_AUTH -w \"/builds/3cixCcwz/0/intern/renovate-runner/renovate/repos/gitlab/extern/007\" docker.io/renovate/composer:1.10.20 bash -l -c \"composer update drenso/symfony-shared --with-dependencies --ignore-platform-reqs --no-ansi --no-interaction --no-scripts --no-autoloader\"","msg":"Executing command","time":"2021-02-25T14:54:08.403Z","v":0}
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":13,"level":20,"logContext":"zecDfDsSF","repository":"extern/007","branch":"renovate/drenso-symfony-shared-1.22.x","cmd":"docker run --rm --name=renovate_composer --label=renovate_child -v \"/builds/3cixCcwz/0/intern/renovate-runner/renovate/repos/gitlab/extern/007\":\"/builds/3cixCcwz/0/intern/renovate-runner/renovate/repos/gitlab/extern/007\" -v \"/builds/3cixCcwz/0/intern/renovate-runner/renovate/cache\":\"/builds/3cixCcwz/0/intern/renovate-runner/renovate/cache\" -e COMPOSER_CACHE_DIR -e COMPOSER_AUTH -w \"/builds/3cixCcwz/0/intern/renovate-runner/renovate/repos/gitlab/extern/007\" docker.io/renovate/composer:1.10.20 bash -l -c \"composer update drenso/symfony-shared --with-dependencies --ignore-platform-reqs --no-ansi --no-interaction --no-scripts --no-autoloader\"","durationMs":57342,"stdout":"","stderr":"Dependency \"symfony/form\" is also a root requirement, but is not explicitly allowed. Ignoring.\nDependency \"twig/twig\" is also a root requirement, but is not explicitly allowed. Ignoring.\nLoading composer repositories with package information\nUpdating dependencies (including require-dev)\nPackage operations: 169 installs, 0 updates, 0 removals\n  - Installing composer/package-versions-deprecated (1.11.99.1): Loading from cache\n  - Installing symfony/flex (v1.12.2): Loading from cache\n  - Installing psr/simple-cache (1.0.1): Loading from cache\n  - Installing psr/http-message (1.0.1): Loading from cache\n  - Installing psr/http-factory (1.0.1): Loading from cache\n  - Installing psr/http-client (1.0.1): Loading from cache\n  - Installing markbaker/matrix (2.1.2): Loading from cache\n  - Installing markbaker/complex (2.0.0): Loading from cache\n  - Installing symfony/polyfill-mbstring (v1.22.1): Loading from cache\n  - Installing myclabs/php-enum (1.8.0): Loading from cache\n  - Installing psr/container (1.0.0): Loading from cache\n  - Installing symfony/service-contracts (v2.2.0): Loading from cache\n  - Installing psr/cache (1.0.1): Loading from cache\n  - Installing symfony/polyfill-php80 (v1.22.1): Loading from cache\n  - Installing symfony/polyfill-php73 (v1.22.1): Loading from cache\n  - Installing symfony/console (v4.4.19): Loading from cache\n  - Installing nikic/php-parser (v4.10.4): Loading from cache\n  - Installing psr/log (1.1.3): Loading from cache\n  - Installing maennchen/zipstream-php (2.1.0): Loading from cache\n  - Installing ezyang/htmlpurifier (v4.13.0): Loading from cache\n  - Installing phpoffice/phpspreadsheet (1.16.1): Loading from cache\n  - Installing twig/twig (v2.14.3): Loading from cache\n  - Installing symfony/process (v4.4.19): Loading from cache\n  - Installing symfony/deprecation-contracts (v2.2.0): Loading from cache\n  - Installing symfony/options-resolver (v4.4.19): Loading from cache\n  - Installing symfony/polyfill-php72 (v1.22.1): Loading from cache\n  - Installing symfony/polyfill-intl-normalizer (v1.22.1): Loading from cache\n  - Installing symfony/polyfill-intl-idn (v1.22.1): Loading from cache\n  - Installing symfony/mime (v5.2.3): Loading from cache\n  - Installing symfony/http-foundation (v5.2.3): Loading from cache\n  - Installing symfony/http-client-contracts (v2.3.1): Loading from cache\n  - Installing symfony/event-dispatcher-contracts (v1.1.9): Loading from cache\n  - Installing symfony/event-dispatcher (v4.4.19): Loading from cache\n  - Installing symfony/var-dumper (v5.2.3): Loading from cache\n  - Installing symfony/debug (v4.4.19): Loading from cache\n  - Installing symfony/error-handler (v4.4.19): Loading from cache\n  - Installing symfony/http-kernel (v4.4.19): Loading from cache\n  - Installing symfony/routing (v4.4.19): Loading from cache\n  - Installing symfony/finder (v4.4.19): Loading from cache\n  - Installing symfony/filesystem (v5.2.3): Loading from cache\n  - Installing symfony/dependency-injection (v5.2.3): Loading from cache\n  - Installing symfony/config (v5.2.3): Loading from cache\n  - Installing symfony/var-exporter (v4.4.19): Loading from cache\n  - Installing symfony/cache-contracts (v2.2.0): Loading from cache\n  - Installing symfony/cache (v4.4.19): Loading from cache\n  - Installing symfony/framework-bundle (v4.4.19): Loading from cache\n  - Installing symfony/polyfill-intl-grapheme (v1.22.1): Loading from cache\n  - Installing symfony/string (v5.2.3): Loading from cache\n  - Installing symfony/inflector (v4.4.19): Loading from cache\n  - Installing symfony/property-access (v4.4.19): Loading from cache\n  - Installing symfony/polyfill-intl-icu (v1.22.1): Loading from cache\n  - Installing symfony/intl (v4.4.19): Loading from cache\n  - Installing symfony/form (v4.4.19): Loading from cache\n  - Installing bobv/latex-bundle (v2.5.1): Loading from cache\n  - Installing kigkonsult/icalcreator (v2.24.2): Loading from cache\n  - Installing bomo/ical-bundle (v1.0.14): Loading from cache\n  - Installing symfony/security (v4.4.19): Loading from cache\n  - Installing symfony/security-bundle (v4.4.19): Loading from cache\n  - Installing phpseclib/phpseclib (2.0.30): Loading from cache\n  - Installing drenso/symfony-oidc-bundle (v1.10.0): Loading from cache\n  - Installing drenso/symfony-shared (1.22.1): Loading from cache\n  - Installing erusev/parsedown (1.7.4): Loading from cache\n  - Installing erusev/parsedown-extra (0.8.1): Loading from cache\n  - Installing react/promise (v2.8.0): Loading from cache\n  - Installing guzzlehttp/streams (3.0.0): Loading from cache\n  - Installing guzzlehttp/ringphp (1.1.1): Loading from cache\n  - Installing elasticsearch/elasticsearch (v6.7.2): Loading from cache\n  - Installing ruflin/elastica (6.1.3): Loading from cache\n  - Installing pagerfanta/pagerfanta (v2.7.0): Loading from cache\n  - Installing friendsofsymfony/elastica-bundle (v5.2.1): Loading from cache\n  - Installing willdurand/jsonp-callback-validator (v1.1.0): Loading from cache\n  - Installing symfony/serializer (v4.4.19): Loading from cache\n  - Installing friendsofsymfony/jsrouting-bundle (2.7.0): Loading from cache\n  - Installing studio-42/elfinder (2.1.57): Loading from cache\n  - Installing symfony/asset (v4.4.19): Loading from cache\n  - Installing symfony/translation-contracts (v2.3.0): Loading from cache\n  - Installing symfony/twig-bridge (v4.4.19): Loading from cache\n  - Installing symfony/twig-bundle (v4.4.19): Loading from cache\n  - Installing helios-ag/fm-elfinder-bundle (10.1): Loading from cache\n  - Installing jms/metadata (2.4.0): Loading from cache\n  - Installing phpstan/phpdoc-parser (0.4.11): Loading from cache\n  - Installing doctrine/lexer (1.2.1): Loading from cache\n  - Installing doctrine/instantiator (1.4.0): Loading from cache\n  - Installing doctrine/annotations (1.12.1): Loading from cache\n  - Installing jms/serializer (3.11.0): Loading from cache\n  - Installing jms/serializer-bundle (3.8.0): Loading from cache\n  - Installing kick-in/exception-handler-bundle (3.0.0): Loading from cache\n  - Installing league/html-to-markdown (4.10.0): Loading from cache\n  - Installing imagine/imagine (1.2.4): Loading from cache\n  - Installing liip/imagine-bundle (2.5.0): Loading from cache\n  - Installing giggsey/locale (1.9): Loading from cache\n  - Installing giggsey/libphonenumber-for-php (8.12.18): Loading from cache\n  - Installing misd/phone-number-bundle (v1.3.2): Loading from cache\n  - Installing php-decimal/stubs (v1.1.0): Loading from cache\n  - Installing php-decimal/php-decimal (v1.1.0): Loading from cache\n  - Installing sensio/framework-extra-bundle (v6.1.1): Loading from cache\n  - Installing symfony/http-client (v4.4.19): Loading from cache\n  - Installing symfony/polyfill-uuid (v1.22.1): Loading from cache\n  - Installing ralouphie/getallheaders (3.0.3): Loading from cache\n  - Installing guzzlehttp/psr7 (1.7.0): Loading from cache\n  - Installing php-http/message-factory (v1.0.2): Loading from cache\n  - Installing clue/stream-filter (v1.5.0): Loading from cache\n  - Installing php-http/message (1.11.0): Loading from cache\n  - Installing php-http/promise (1.1.0): Loading from cache\n  - Installing php-http/httplug (2.2.0): Loading from cache\n  - Installing php-http/discovery (1.13.0): Loading from cache\n  - Installing php-http/client-common (2.3.0): Loading from cache\n  - Installing guzzlehttp/promises (1.4.0): Loading from cache\n  - Installing jean85/pretty-package-versions (1.6.0): Loading from cache\n  - Installing sentry/sentry (2.5.2): Loading from cache\n  - Installing http-interop/http-factory-guzzle (1.0.0): Loading from cache\n  - Installing sentry/sdk (2.2.0)\n  - Installing sentry/sentry-symfony (3.5.3): Loading from cache\n  - Installing doctrine/event-manager (1.1.1): Loading from cache\n  - Installing doctrine/reflection (1.2.2): Loading from cache\n  - Installing doctrine/collections (1.6.7): Loading from cache\n  - Installing doctrine/cache (1.10.2): Loading from cache\n  - Installing doctrine/persistence (1.3.8): Loading from cache\n  - Installing doctrine/inflector (1.4.3): Loading from cache\n  - Installing doctrine/common (2.13.3): Loading from cache\n  - Installing behat/transliterator (v1.3.0): Loading from cache\n  - Installing gedmo/doctrine-extensions (v3.0.3): Loading from cache\n  - Installing stof/doctrine-extensions-bundle (v1.5.0): Loading from cache\n  - Installing symfony/dotenv (v4.4.19): Loading from cache\n  - Installing symfony/ldap (v4.4.19): Loading from cache\n  - Installing symfony/lock (v4.4.19): Loading from cache\n  - Installing egulias/email-validator (2.1.25): Loading from cache\n  - Installing symfony/mailer (v4.4.19): Loading from cache\n  - Installing symfony/messenger (v4.4.19): Loading from cache\n  - Installing monolog/monolog (1.26.0): Loading from cache\n  - Installing symfony/monolog-bridge (v4.4.19): Loading from cache\n  - Installing symfony/monolog-bundle (v3.6.0): Loading from cache\n  - Installing symfony/stopwatch (v4.4.19): Loading from cache\n  - Installing laminas/laminas-zendframework-bridge (1.1.1): Loading from cache\n  - Installing laminas/laminas-eventmanager (3.3.0): Loading from cache\n  - Installing laminas/laminas-code (3.4.1): Loading from cache\n  - Installing friendsofphp/proxy-manager-lts (v1.0.3): Loading from cache\n  - Installing doctrine/dbal (2.12.1): Loading from cache\n  - Installing doctrine/migrations (2.3.2): Loading from cache\n  - Installing symfony/doctrine-bridge (v4.4.19): Loading from cache\n  - Installing doctrine/sql-formatter (1.1.1): Loading from cache\n  - Installing doctrine/doctrine-bundle (2.2.3): Loading from cache\n  - Installing doctrine/doctrine-migrations-bundle (2.2.2): Loading from cache\n  - Installing doctrine/orm (2.7.5): Loading from cache\n  - Installing symfony/orm-pack (v1.2.0): Loading from cache\n  - Installing symfony/templating (v4.4.19): Loading from cache\n  - Installing symfony/translation (v4.4.19): Loading from cache\n  - Installing symfony/validator (v4.4.19): Loading from cache\n  - Installing symfony/webpack-encore-bundle (v1.11.1): Loading from cache\n  - Installing symfony/yaml (v4.4.19): Loading from cache\n  - Installing symfony/css-selector (v4.4.19): Loading from cache\n  - Installing tijsverkoyen/css-to-inline-styles (2.2.3): Loading from cache\n  - Installing twig/cssinliner-extra (v3.3.0): Loading from cache\n  - Installing twig/extra-bundle (v3.3.0): Loading from cache\n  - Installing lorenzo/pinky (1.0.5): Loading from cache\n  - Installing twig/inky-extra (v3.3.0): Loading from cache\n  - Installing twig/intl-extra (v3.3.0): Loading from cache\n  - Installing vich/uploader-bundle (1.16.1): Loading from cache\n  - Installing dama/doctrine-test-bundle (v6.5.0): Loading from cache\n  - Installing doctrine/data-fixtures (1.5.0): Loading from cache\n  - Installing doctrine/doctrine-fixtures-bundle (3.4.0): Loading from cache\n  - Installing drenso/phan-extensions (v3.2.1): Loading from cache\n  - Installing symfony/maker-bundle (v1.29.1): Loading from cache\n  - Installing symfony/web-profiler-bundle (v4.4.19): Loading from cache\n  - Installing symfony/profiler-pack (v1.0.5): Loading from cache\n  - Installing symfony/phpunit-bridge (v5.2.3): Loading from cache\n  - Installing symfony/dom-crawler (v4.4.19): Loading from cache\n  - Installing symfony/browser-kit (v4.4.19): Loading from cache\n  - Installing symfony/test-pack (v1.0.7): Loading from cache\nPackage guzzlehttp/streams is abandoned, you should avoid using it. No replacement was suggested.\nPackage guzzlehttp/ringphp is abandoned, you should avoid using it. No replacement was suggested.\nPackage misd/phone-number-bundle is abandoned, you should avoid using it. Use odolbeau/phone-number-bundle instead.\nPackage doctrine/reflection is abandoned, you should avoid using it. Use roave/better-reflection instead.\nWriting lock file\n104 packages you are using are looking for funding.\nUse the `composer fund` command to find out more!\n","msg":"exec completed","time":"2021-02-25T14:55:05.683Z","v":0}
```

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/master/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] This is a really small bug, it does not need a reproduction (think small typo)
- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

This relates in some level to https://github.com/renovatebot/renovate/issues/3558, although that issue also talks about updating the flex constraint.

I'm tagging @nicolas-grekas here as well, as he might consider this as a flex and/or composer bug. Although this could well be a known limitation of composer plugins.

### PR DESCRIPTION
**Pull request by bobvandevijver**
## Changes:

This PR adds three new configuration options: 

- `installBeforeUpdate`: Configuration option, used to install the dependencies before trying to run the update task
- `allowPlugins`: Self hosted option, can be used to enable plugin execution
- `ignorePlugins`: Configuration option, to override the global `allowPlugins` options when it results in problems

## Context:

This fixes #8866, as when these options are configured correctly (see the example below) the Symfony Flex Composer plugin is now installed before the update task is run, so it can influence the update process. 
This PR also extends the work done in #10989 as the `--no-plugins` Composer behaviour is now configurable, separately from the `--no-scripts` option with similar options. 

The options have only been implemented for the Composer manager at this time, but they can probably be extended to other managers in the future whenever needed.

Correct options for Symfony Flex users:
- `installBeforeUpdate` set to `true`
- `allowPlugins`  set to `true`
- `ignorePlugins` omitted or set to `false`

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

I've run this on the original reproduction repository for #8886, which resulted in https://github.com/bobvandevijver/renovate-repro-8866/pull/2. The updated lock file can be inspected to see that the symfony packages are now kept in the correct range, which was not the case when Renovate without this patch was run (https://github.com/bobvandevijver/renovate-repro-8866/blob/master/reproduction-logs/invalid.patch). 

Most interesting debug log line is this one, which includes the magic `Restricting packages listed in "symfony/symfony" to "4.4.*"` line:
```
{"name":"renovate","hostname":"XPS15-BOB","pid":3364,"level":20,"logContext":"CJTQJmnr0","repository":"bobvandevijver/renovate-repro-8866","branch":"renovate/drenso-symfony-shared-1.x","cmd":"composer update drenso/symfony-shared --with-dependencies --ignore-platform-reqs --no-ansi --no-interaction --no-scripts --no-autoloader","durationMs":8044,"stdout":"","stderr":"Loading composer repositories with package information\nRestricting packages listed in \"symfony/symfony\" to \"4.4.*\"\nUpdating dependencies\nLock file operations: 1 install, 29 updates, 0 removals\n  - Upgrading drenso/symfony-shared (1.22.0 => 1.24.0)\n  - Upgrading psr/container (1.0.0 => 1.1.1)\n  - Upgrading psr/log (1.1.3 => 2.0.0)\n  - Upgrading symfony/config (v4.4.19 => v4.4.30)\n  - Upgrading symfony/debug (v4.4.19 => v4.4.31)\n  - Upgrading symfony/dependency-injection (v4.4.19 => v4.4.31)\n  - Upgrading symfony/error-handler (v4.4.19 => v4.4.30)\n  - Upgrading symfony/event-dispatcher (v4.4.19 => v4.4.30)\n  - Upgrading symfony/filesystem (v4.4.19 => v4.4.27)\n  - Upgrading symfony/form (v4.4.19 => v4.4.31)\n  - Upgrading symfony/http-client-contracts (v2.3.1 => v2.4.0)\n  - Upgrading symfony/http-foundation (v4.4.19 => v4.4.30)\n  - Upgrading symfony/http-kernel (v4.4.19 => v4.4.32)\n  - Upgrading symfony/inflector (v4.4.19 => v4.4.27)\n  - Upgrading symfony/intl (v4.4.19 => v4.4.31)\n  - Upgrading symfony/mime (v4.4.19 => v4.4.31)\n  - Upgrading symfony/options-resolver (v4.4.19 => v4.4.30)\n  - Upgrading symfony/polyfill-ctype (v1.22.1 => v1.23.0)\n  - Upgrading symfony/polyfill-intl-icu (v1.22.1 => v1.23.0)\n  - Upgrading symfony/polyfill-intl-idn (v1.22.1 => v1.23.0)\n  - Upgrading symfony/polyfill-intl-normalizer (v1.22.1 => v1.23.0)\n  - Upgrading symfony/polyfill-mbstring (v1.22.1 => v1.23.1)\n  - Upgrading symfony/polyfill-php72 (v1.22.1 => v1.23.0)\n  - Upgrading symfony/polyfill-php73 (v1.22.1 => v1.23.0)\n  - Upgrading symfony/polyfill-php80 (v1.22.1 => v1.23.1)\n  - Locking symfony/polyfill-php81 (v1.23.0)\n  - Upgrading symfony/property-access (v4.4.19 => v4.4.30)\n  - Upgrading symfony/service-contracts (v2.2.0 => v2.4.0)\n  - Upgrading symfony/var-dumper (v4.4.19 => v4.4.31)\n  - Upgrading twig/twig (v3.3.0 => v3.3.3)\nWriting lock file\nInstalling dependencies from lock file (including require-dev)\nPackage operations: 1 install, 29 updates, 0 removals\n  - Downloading symfony/polyfill-mbstring (v1.23.1)\n  - Downloading symfony/polyfill-ctype (v1.23.0)\n  - Downloading twig/twig (v3.3.3)\n  - Downloading symfony/polyfill-php80 (v1.23.1)\n  - Downloading symfony/polyfill-php73 (v1.23.0)\n  - Downloading symfony/polyfill-php72 (v1.23.0)\n  - Downloading symfony/polyfill-intl-normalizer (v1.23.0)\n  - Downloading symfony/polyfill-intl-idn (v1.23.0)\n  - Downloading symfony/mime (v4.4.31)\n  - Downloading symfony/http-foundation (v4.4.30)\n  - Downloading symfony/http-client-contracts (v2.4.0)\n  - Downloading symfony/event-dispatcher (v4.4.30)\n  - Downloading symfony/var-dumper (v4.4.31)\n  - Downloading psr/log (2.0.0)\n  - Downloading symfony/debug (v4.4.31)\n  - Downloading symfony/error-handler (v4.4.30)\n  - Downloading symfony/http-kernel (v4.4.32)\n  - Downloading psr/container (1.1.1)\n  - Downloading symfony/service-contracts (v2.4.0)\n  - Downloading symfony/inflector (v4.4.27)\n  - Downloading symfony/property-access (v4.4.30)\n  - Downloading symfony/options-resolver (v4.4.30)\n  - Downloading symfony/polyfill-intl-icu (v1.23.0)\n  - Downloading symfony/intl (v4.4.31)\n  - Downloading symfony/form (v4.4.31)\n  - Downloading symfony/dependency-injection (v4.4.31)\n  - Downloading symfony/polyfill-php81 (v1.23.0)\n  - Downloading symfony/filesystem (v4.4.27)\n  - Downloading symfony/config (v4.4.30)\n  - Downloading drenso/symfony-shared (1.24.0)\n  - Upgrading symfony/polyfill-mbstring (v1.22.1 => v1.23.1): Extracting archive\n  - Upgrading symfony/polyfill-ctype (v1.22.1 => v1.23.0): Extracting archive\n  - Upgrading twig/twig (v3.3.0 => v3.3.3): Extracting archive\n  - Upgrading symfony/polyfill-php80 (v1.22.1 => v1.23.1): Extracting archive\n  - Upgrading symfony/polyfill-php73 (v1.22.1 => v1.23.0): Extracting archive\n  - Upgrading symfony/polyfill-php72 (v1.22.1 => v1.23.0): Extracting archive\n  - Upgrading symfony/polyfill-intl-normalizer (v1.22.1 => v1.23.0): Extracting archive\n  - Upgrading symfony/polyfill-intl-idn (v1.22.1 => v1.23.0): Extracting archive\n  - Upgrading symfony/mime (v4.4.19 => v4.4.31): Extracting archive\n  - Upgrading symfony/http-foundation (v4.4.19 => v4.4.30): Extracting archive\n  - Upgrading symfony/http-client-contracts (v2.3.1 => v2.4.0): Extracting archive\n  - Upgrading symfony/event-dispatcher (v4.4.19 => v4.4.30): Extracting archive\n  - Upgrading symfony/var-dumper (v4.4.19 => v4.4.31): Extracting archive\n  - Upgrading psr/log (1.1.3 => 2.0.0): Extracting archive\n  - Upgrading symfony/debug (v4.4.19 => v4.4.31): Extracting archive\n  - Upgrading symfony/error-handler (v4.4.19 => v4.4.30): Extracting archive\n  - Upgrading symfony/http-kernel (v4.4.19 => v4.4.32): Extracting archive\n  - Upgrading psr/container (1.0.0 => 1.1.1): Extracting archive\n  - Upgrading symfony/service-contracts (v2.2.0 => v2.4.0): Extracting archive\n  - Upgrading symfony/inflector (v4.4.19 => v4.4.27): Extracting archive\n  - Upgrading symfony/property-access (v4.4.19 => v4.4.30): Extracting archive\n  - Upgrading symfony/options-resolver (v4.4.19 => v4.4.30): Extracting archive\n  - Upgrading symfony/polyfill-intl-icu (v1.22.1 => v1.23.0): Extracting archive\n  - Upgrading symfony/intl (v4.4.19 => v4.4.31): Extracting archive\n  - Upgrading symfony/form (v4.4.19 => v4.4.31): Extracting archive\n  - Upgrading symfony/dependency-injection (v4.4.19 => v4.4.31): Extracting archive\n  - Installing symfony/polyfill-php81 (v1.23.0): Extracting archive\n  - Upgrading symfony/filesystem (v4.4.19 => v4.4.27): Extracting archive\n  - Upgrading symfony/config (v4.4.19 => v4.4.30): Extracting archive\n  - Upgrading drenso/symfony-shared (1.22.0 => 1.24.0): Extracting archive\nPackage symfony/inflector is abandoned, you should avoid using it. Use use `EnglishInflector` from the String component instead instead.\n30 packages you are using are looking for funding.\nUse the `composer fund` command to find out more!\n\nWhat about running composer global require symfony/thanks && composer thanks now?\nThis will spread some 💖  by sending a ★  to the GitHub repositories of your fellow package maintainers.\n\nRun composer recipes at any time to see the status of your Symfony recipes.\n\n","msg":"exec completed","time":"2021-10-02T19:20:03.695Z","v":0}
```

### CODE REVIEW

**Comment by bobvandevijver** [comment]
I removed the `installBeforeUpdate` option in favorite of automatic Symfony Flex detection. Note that I did already merge PR #12185 in here, as that is a very needed and welcome optimization for this change.

File: docs/usage/configuration-options.md [hunk 1]
```diff
@@ -1089,6 +1089,10 @@ For instance if you have a project with an `"examples/"` directory you wish to i
 
 Useful to know: Renovate's default ignore is `node_modules` and `bower_components` only, however if you are extending the popular `config:base` preset then it adds ignore patterns for `vendor`, `examples`, `test(s)` and `fixtures` directories too.
 
+## ignorePlugins
+
+Applicable for Composer only for now. Set this to `true` if running plugins causes problems.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:1094]:
```suggestion
Set this to `true` if running plugins causes problems.
Applicable for Composer only for now.
```

File: lib/manager/composer/artifacts.ts [hunk 2]
```diff
@@ -73,6 +73,30 @@ function getAuthJson(): string | null {
...
+      });
+    }
+  }
+
+  args += ' --no-ansi --no-interaction';
+  if (!getGlobalConfig().allowScripts || config.ignoreScripts) {
+    args += ' --no-scripts --no-autoloader';
+  }
+
+  if (!getGlobalConfig().allowPlugins || config.ignorePlugins) {
```
Comment (rarkins) [inline lib/manager/composer/artifacts.ts:93]:
```suggestion
  if (getGlobalConfig().allowPlugins && !config.ignorePlugins) {
```

bit easier to understand

**Review by rarkins — DISMISSED** [review]

**Comment by bobvandevijver** [comment]
Hmm, I will check the tests tomorrow!

File: lib/manager/composer/artifacts.ts (hunk 2 above)
Comment (bobvandevijver) [reply lib/manager/composer/artifacts.ts]: I misread this yesterday, but unfortunately this changes the behaviour as we need to disable the plugins when not allowed, instead of the other way around. This suggestion would need to be wrapped with a negation, making it less understandable. This notation was already used by the way: https://github.com/renovatebot/renovate/pull/11990/files#diff-d5d778f4c4a0495c698f14938cd1deb6da79f83e62ff015909a705ef1a0f9319L142

File: lib/manager/composer/artifacts.spec.ts [hunk 3]
```diff
@@ -367,4 +368,89 @@ describe('manager/composer/artifacts', () => {
...
     expect(execSnapshots).toMatchSnapshot();
   });
+
+  it('installs before running the update when symfony flex is installed', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(
+      '{"packages":[{"name":"symfony/flex","version":"1.17.1"}]}'
+    );
+    const execSnapshots = mockExecAll(exec);
+    fs.readLocalFile.mockResolvedValueOnce('{ }');
+    git.getRepoStatus.mockResolvedValue({
```
Comment (viceice) [inline lib/manager/composer/artifacts.spec.ts:378]: Should also use once?

File: lib/manager/composer/artifacts.ts [hunk 4]
```diff
@@ -129,21 +165,11 @@ export async function updateArtifacts({
           'update ' + updatedDeps.map((dep) => quote(dep.depName)).join(' ')
         ).trim() + ' --with-dependencies';
     }
-    if (config.composerIgnorePlatformReqs) {
```
Comment (viceice) [inline lib/manager/composer/artifacts.ts:132]: Can you extract the refactor to a separate PR to merge first please.

File: lib/manager/composer/artifacts.ts [hunk 5]
```diff
@@ -73,6 +73,30 @@ function getAuthJson(): string | null {
   return is.emptyObject(authJson) ? null : JSON.stringify(authJson);
 }
 
+function getComposerArguments(config): string {
```
Comment (viceice) [inline lib/manager/composer/artifacts.ts:76]: Better move this to a `./utils.ts` file as refactor PR, see other comment.

Comment (bobvandevijver) [reply lib/manager/composer/artifacts.ts:76]: Done, see #12443.

File: lib/manager/composer/artifacts.ts (hunk 4 above)
Comment (bobvandevijver) [reply lib/manager/composer/artifacts.ts:132]: Done, see #12443.

**Comment by bobvandevijver** [comment]
Will rebase this (and check the other comment) when #12443 has been processed 👍🏻

File: lib/manager/composer/artifacts.spec.ts (hunk 3 above)
Comment (bobvandevijver) [reply lib/manager/composer/artifacts.spec.ts]: That works as well, I took the liberty to change all occurrences in the `artifacts.spec` file.

File: lib/manager/composer/artifacts.ts [hunk 6]
```diff
@@ -119,6 +117,20 @@ export async function updateArtifacts({
         tagScheme: composerVersioningId,
       },
     };
+
+    // Determine whether Symfony Flex has been installed
+    const hasSymfonyFlex =
+      existingLockFile.packages?.some((p) => p.name === 'symfony/flex') ||
+      existingLockFile['packages-dev']?.some((p) => p.name === 'symfony/flex');
```
Comment (viceice) [inline lib/manager/composer/artifacts.ts:124]: Can you move this to a util function and make it more generic, so it's easily extendable to more packages in future if requirewd?

Comment (bobvandevijver) [reply lib/manager/composer/artifacts.ts:124]: Something like this?

```js
export function hasComposerDependencyInstalled(
  lock: ComposerLock,
  dependency: string,
  includeDev?: boolean,
): boolean {
  if (lock.packages?.some((p) => p.name === dependency)) {
    return true;
  }

  return includeDev && lock['packages-dev']?.some((p) => p.name === dependency);
}
```

File: lib/manager/composer/utils.ts [hunk 7]
```diff
@@ -79,6 +83,21 @@ export function getPhpConstraint(constraints: Record<string, string>): string {
...
+): boolean {
+  if (lock.packages?.some((p) => p.name === dependency)) {
+    return true;
+  }
+
+  return (
+    includeDev === true &&
+    lock['packages-dev']?.some((p) => p.name === dependency) === true
+  );
+}
```
Comment (viceice) [inline lib/manager/composer/utils.ts:99]: I think about something like this:

```suggestion

const depRequireInstall = new Set(['symfony/flex']);

export function requireComposerDependencyInstalled(
  lock: ComposerLock
): boolean {
  return (
    lock.packages?.some((p) => depRequireInstall.has(p.name)) === true ||
    lock['packages-dev']?.some((p) => depRequireInstall.has(p.name)) === true
  );
}
```

Comment (bobvandevijver) [reply lib/manager/composer/utils.ts:99]: Okay, see latest update.

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
> Okay, see latest update.

Next time please simple avoid those simple comments and use the re-request review button. So we get less noise 😉

**Comment by bobvandevijver** [comment]
@rarkins Your review is still being marked as pending, but that is not true. I'm not sure why Github thinks it is though...

**Review by viceice — DISMISSED** [review]

**Review by rarkins — DISMISSED** [review]

**Review by rarkins — APPROVED** [review]
